core/commonMain/src/ArgParser.kt

```diff
@@ -342,24 +371,117 @@ open class ArgParser(
+        val argValue = if (prefixStyle == OptionPrefixStyle.GNU) null else options[optionString]
+        if (argValue != null) {
+            saveStandardOptionForm(argValue, argIterator)
+            return true
+        } else {
+            // Check GNU style of options.
+            if (prefixStyle == OptionPrefixStyle.GNU) {
+                // Option without a parameter.
+                if (options[optionString]?.descriptor?.type?.hasParameter == false) {
+                    saveOptionWithoutParameter(options[optionString]!!)
+                    return true
+                }
+                // Option with parameters.
+                val optionParts = optionString.split('=', limit = 2)
+                if (optionParts.size != 2)
+                    return false
+                if (options[optionParts[0]] != null) {
+                    saveAsOption(options[optionParts[0]]!!, optionParts[1])
+                    return true
+                }
+            }
+        }
+        return false
+    }
 
     /**
-     * Try to recognize command line element as short form of option.
+     * Save option without parameter.
+     *
+     * @param argValue argument value with all information about option.
+     */
+    private fun saveOptionWithoutParameter(argValue: ParsingValue<*, *>) {
+        // Boolean flags.
+        if (argValue.descriptor.fullName == "help") {
+            println(makeUsage())
+            exitProcess(0)
+        }
+        saveAsOption(argValue, "true")
+    }
+
+    /**
+     * Save option described with standard separated form `--name value`.
+     *
+     * @param argValue argument value with all information about option.
+     * @param argIterator iterator over command line arguments.
+     */
+    private fun saveStandardOptionForm(argValue: ParsingValue<*, *>, argIterator: Iterator<String>) {
+        if (argValue.descriptor.type.hasParameter) {
+            if (argIterator.hasNext()) {
+                saveAsOption(argValue, argIterator.next())
+            } else {
+                // An error, option with value without value.
+                printError("No value for ${argValue.descriptor.textDescription}")
+            }
+        } else {
+            saveOptionWithoutParameter(argValue)
+        }
+    }
+
+    /**
+     * Try to recognize and save command line element as short form of option.
      *
      * @param candidate string with candidate in options.
+     * @param argIterator iterator over command line arguments.
      */
-    private fun recognizeOptionShortForm(candidate: String) =
-            if (candidate.startsWith(optionShortFromPrefix))
-                shortNames[candidate.substring(optionShortFromPrefix.length)]
-            else null
+    private fun recognizeAndSaveOptionShortForm(candidate: String, argIterator: Iterator<String>): Boolean {
+        if (!candidate.startsWith(optionShortFromPrefix)) return false
```

Comment:
Wouldn't be it a problem, that when a candidate starts with the long prefix, it also starts with the short prefix?
When these two prefixes are distinct we seem to be doing an excess work here that would never succeed.
In other words, we should not try to parse `--option` string in this method.